### ISSUE TITLE
Feature: Suppressing rule violations

### ISSUE THREAD

**Issue by carlpett** [open-policy-agent/conftest#315]
I posted this on Slack a while ago, but after running with the workaround-ish "solution" for a while, I think it might be worth discussing if conftest should have some sort of pattern for ignoring/suppressing violations.

As I guess many others, we're using conftest as part of our linter/ci-check process for Kubernetes manifests. As we're getting further in codifying our best practices, we're starting to see where we have things which we want to be true in 90% of the cases, but there are valid exceptions. For example, while we almost never want containers to run as root, there are a handful of valid cases where it needs to be allowed.

What we've been doing is basically throwing an annotation on the Kubernetes resource like for example adding `check-ignore: run-as-nonroot` if that particular deployment would be okay running as root.
While this _works_, it is not very nice to work with for a couple of reasons:
- We need to not forget to add the "check if ignored" on all rules
- We need to ensure the proper "id" for suppression is output in the `msg`, for discoverability (and if it would change, also change the msg)
- We end up with linter annotation in production manifests (mostly cosmetic, but not desirable)
- Understanding which manifests use suppressions/ignores requires trawling through them all
A suggestion in the discussions on Slack was using data files instead, which would address the last two points to some extent, but not the first two.

Looking at other linters, the two common patterns would be a config file (for example golangci-lint does this), or magic comments. I think the comments path might be difficult here since policies don't always cleanly apply to a single line, or even a single file.

I don't have a clear idea of how this could be implemented, yet. But I would guess that making this a first class feature would likely require some changes so the rules have identites, at least?
There might also be cases which the policy writer might not accept being ignored? Is this something that would need to be handled?

**Comment by Blokje5** [comment]
I agree this would be a great feature! It is something that seems to come up time and time again. It makes sense to see if there is a way to include first class support for surpressions. There are some complications however:

1. Multiple `deny` rules in `Rego` is essentially a single incrementally defined `deny` rule. That means if your policies include everything in a single `deny` rule it is hard to differentiate between the different incremental definitions. Luckily conftest can query rules of the form `^(deny|violation)(_[a-zA-Z0-9]+)*$`. That does potentially mean that incremental definitions would not work well with surpressions, but I guess that is not a big issue.

2. Differentiating between configuration objects could be hard. I think most people are looking for a way to make exceptions for specific instances (e.g. a specific deployment in Kubernetes) but inside conftest we do not have that information available. E.g. we do not read the name value from inside the deployment spec. I don't think it would be a great user experience if we would only allow surpressions based on the name of the configuration file.

Whatever approach we choose should at least be flexible in how exceptions are configured. Some users would want surpressions based on instance names, some might want to filter on labels and I think both are valid approaches. Plus conftest is also commonly used for validation of other configuration file so the approach has to be agnostic to what type of configuration file it is checking (e.g. it should also work for Terraform).

For that reason, I think one possible approach would be to use Rego policies to filter the exceptions. Of course this is quite similar to the workaround people are often using now, but we could integrate a standard filter (pre or post query execution) in Rego. 

e.g.

```
package exceptions

filter[rules] {
  input.kind = "Deployment"
  input.metadata.name = "ExceptionInstance"

  rules = ["deny_root_user"]
}
```

This could give us a mapping from configuration file to rules. We can then filter out the rules that this configuration file is excluded from. However, this does mean the filtering would filter out the entire configuration file when using the combine flag, which I don't think is the expected behaviour. But maybe it is ok to document that.

As a starting point for further discussion I would like to get a clear idea of what is wanted by the community. So far I think the following requirements make sense:

1. Suppressing rule violations should be flexible and it should be possible to base it on internal information (e.g. Deployment name, Specific annotations).
2. Suppressing rule violations should be agnostic of the configuration file type. No code should be needed to check if a configuration file is e.g. a Kubernetes deployment in order for the surpressions to work.
3. It should be possible to suppress specific rules. Incremental definitions will probably end up being counted as a single rule here.
4. It should be possible to see which rule surpressions where triggered.

**Comment by carlpett** [comment]
Great ideas! If I understand you correctly, this:
```rego
filter[rules] {
  [... some logic ...]

  rules = ["some-rule"]
}
```
... would mean that any input matched by the predicates would **not** have `some-rule` considered a violation? And we could then potentially stack these, so if the same input **also** matched this filter:
```rego
filter[rules] {
  [... some logic ...]

  rules = ["other-rule"]
}
```
... it'd ignore violations of `["some-rule","other-rule"]`? Seems like a very reasonable and flexible solution!

---

As an "acceptance test", basically we'd expect to write something like this:

Policy:
```rego
deny_run_as_root[msg] {
  kubernetes.is_deployment
  not input.spec.template.spec.securityContext.runAsNonRoot

  msg = sprintf("Containers must not run as root in Deployment %s", [name])
}
```
Exception:
```rego
filter[rules] {
  input.kind = "Deployment"
  input.metadata.name = "this-can-run-as-root"

  rules = ["run_as_root"]
}
```
Input:
```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: this-cannot-run-as-root
spec:
  template:
    spec:
      containers:
      - name: root-container
        image: nginx
        ports:
        - containerPort: 8080

---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: this-can-run-as-root
spec:
  template:
    spec:
      containers:
      - name: root-container
        image: nginx
        ports:
        - containerPort: 8080
```
... and given all that, we'd get a single error, `Containers must not run as root in Deployment this-cannot-run-as-root`, while the other deployment would be fine.

---

To get a bit nitty on the details, naming-wise we now have three different terms in just this thread - suppressions, filters and exceptions. Unless there is something even more general possible with this mechanism, `filter` seems a bit generic, and could even be interpreted as inclusion or exclusion? Exceptions on the other hand are (at least to me) very clear, eg `exception[rules] { ... }`.

From my point of view, your requirement list makes a lot of sense, and should be complete enough to make a first implementation from. It'll probably shake out a few things to think about.

**Comment by Blokje5** [comment]
@carlpett Thanks for providing some feedback! I think it is indeed a good point that we should be clear on the naming. Let's stick to exceptions from now on. If you want to help out implementing this, I am more then happy to provide you with some pointers how to get started and where to look. Alternatively I can also give it a go.

**Comment by carlpett** [comment]
@Blokje5 I actually started and have a working prototype. Would the best approach be to submit a PR on what I've got and go from there, perhaps? I'd expect there to be better ways to go about it than what I've got :)

**Comment by Blokje5** [comment]
Closed by #334
